plugin/metrics/prometheus/metricsstore/reader.go

```diff
@@ -16,40 +16,68 @@ package metricsstore
 
 import (
 	"context"
+	"net"
+	"net/http"
 	"time"
 
+	"github.com/prometheus/client_golang/api"
+	promapi "github.com/prometheus/client_golang/api/prometheus/v1"
 	"go.uber.org/zap"
 
 	"github.com/jaegertracing/jaeger/proto-gen/api_v2/metrics"
 	"github.com/jaegertracing/jaeger/storage/metricsstore"
 )
 
 // MetricsReader is a Prometheus metrics reader.
-type MetricsReader struct{}
+type MetricsReader struct {
+	client promapi.API
+	logger *zap.Logger
+}
 
-// NewMetricsReader returns a new MetricsReader, assigning the first reachable host:port from the provided list.
-// This host:port forms part of the URL to call when making queries to the underlying metrics store.
-func NewMetricsReader(logger *zap.Logger, hostPorts []string, connTimeout time.Duration) (*MetricsReader, error) {
-	// TODO: Implement me
-	return &MetricsReader{}, nil
+// NewMetricsReader returns a new MetricsReader.
+func NewMetricsReader(logger *zap.Logger, hostPort string, connectTimeout time.Duration) (*MetricsReader, error) {
+	// KeepAlive and TLSHandshake timeouts are kept to existing Prometheus client's
+	// DefaultRoundTripper to simplify user configuration and may be made configurable when required.
+	roundTripper := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   connectTimeout,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		TLSHandshakeTimeout: 10 * time.Second,
+	}
+
+	client, err := api.NewClient(api.Config{
+		Address:      "http://" + hostPort,
+		RoundTripper: roundTripper,
+	})
+	if err != nil {
+		return nil, err
+	}
+	mr := &MetricsReader{
+		client: promapi.NewAPI(client),
+		logger: logger,
+	}
+	logger.Info("Prometheus reader initialized", zap.String("addr", hostPort))
+	return mr, nil
 }
 
 // GetLatencies gets the latency metrics for the given set of latency query parameters.
-func (m *MetricsReader) GetLatencies(ctx context.Context, params *metricsstore.LatenciesQueryParameters) ([]metrics.Metric, error) {
+func (m *MetricsReader) GetLatencies(ctx context.Context, params *metricsstore.LatenciesQueryParameters) (metrics.MetricFamily, error) {
 	// TODO: Implement me
-	return nil, nil
+	return metrics.MetricFamily{}, nil
 }
 
 // GetCallRates gets the call rate metrics for the given set of call rate query parameters.
-func (m *MetricsReader) GetCallRates(ctx context.Context, params *metricsstore.CallRateQueryParameters) ([]metrics.Metric, error) {
+func (m *MetricsReader) GetCallRates(ctx context.Context, params *metricsstore.CallRateQueryParameters) (metrics.MetricFamily, error) {
 	// TODO: Implement me
-	return nil, nil
+	return metrics.MetricFamily{}, nil
```

Comment:
Any reason to return MetricFamily by value instead of a pointer?

Reply:
No strong reason, it's just a continuation from the API definition which I defined as not nullable.

Perhaps it's better to pass by pointer for performance reasons?

Reply:
and the reason for defining `MetricFamily` as not nullable is to act as a forcing-function to return an empty result for both cases where there is no data in the result as well as an error, which I think is more meaningful (as it also contains meta-data), rather than `nil`.